packages/models-library/src/models_library/progress_bar.py

```diff
@@ -6,17 +6,72 @@
 ProgressUnit: TypeAlias = Literal["Byte"]
 
 
+class ProgressStructuredMessage(BaseModel):
+    description: str
+    current: float
+    total: int
+    unit: str | None
+    sub: "ProgressStructuredMessage | None"
+
+    class Config:
+        schema_extra: ClassVar[dict[str, Any]] = {
+            "examples": [
+                {
+                    "description": "some description",
+                    "current": 12.2,
+                    "total": 123,
+                },
+                {
+                    "description": "some description",
+                    "current": 12.2,
+                    "total": 123,
+                    "unit": "Byte",
+                },
+                {
+                    "description": "downloading",
+                    "current": 2.0,
+                    "total": 5,
+                    "sub": {
+                        "description": "port 2",
+                        "current": 12.2,
+                        "total": 123,
+                        "unit": "Byte",
+                    },
+                },
+            ]
+        }
+
+
 class ProgressReport(BaseModel):
     actual_value: float
     total: float
     unit: ProgressUnit | None = None
+    message: ProgressStructuredMessage | None = None
 
     @property
     def percent_value(self) -> float:
         if self.total != 0:
             return max(min(self.actual_value / self.total, 1.0), 0.0)
         return 0
 
+    def _recursive_compose_message(self, struct_msg: ProgressStructuredMessage) -> str:
+        msg = f"{struct_msg.description}"
+        if struct_msg.sub:
+            return f"{msg}/{self._recursive_compose_message(struct_msg.sub)}"
+        msg = f"{msg} {struct_msg.current} / {struct_msg.total}"
+        return f"{msg} {struct_msg.unit}" if struct_msg.unit else f"{msg} %"
```

Comment:
so, since I am not giving away % no it cannot be %.
I currently only have units that are useful, which currently is no unit or Byte. if there are other later, they can be added.